_posts/2020-04-09-flink-serialization-tuning-vol-1.md

```diff
@@ -0,0 +1,277 @@
+
+```java
+env.getConfig().enableForceAvro();
+```
+
+## Kryo
+
+Any class or object which does not fall into the categories above or is covered by a Flink-provided special serializer is de/serialized with a fallback to [Kryo](https://github.com/EsotericSoftware/kryo) (currently version 2.24.0) which is a powerful and generic serialization framework in Java. Flink calls such a type a *generic type* and you may stumble upon `GenericTypeInfo` when debugging code. If you are using Kryo serialization, make sure to register your types with kryo:
+
+```java
+env.getConfig().registerKryoType(MyCustomType.class);
+```
+Registering types adds them to an internal map of classes to tags so that, during serialization, Kryo does not have to add the fully qualified class names as a prefix into the serialized form. Instead, Kryo uses these (integer) tags to identify the underlying classes and reduce serialization overhead.
+
+<div class="alert alert-info" markdown="1">
+<span class="label label-info" style="display: inline-block"><span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span> Note</span>
+Flink will store Kryo serializer mappings from type registrations in its checkpoints and savepoints and will retain them across job (re)starts.
+</div>
+
+### Disabling Kryo
+
+If desired, you can disable the Kryo fallback, i.e. the ability to serialize generic types, by calling
+
+```java
+env.getConfig().disableGenericTypes();
+```
+
+This is mostly useful for finding out where these fallbacks are applied and replacing them with better serializers. If your job has any generic types with this configuration, it will fail with
+
+> Exception in thread "main" java.lang.UnsupportedOperationException: Generic types have been disabled in the ExecutionConfig and type … is treated as a generic type.
+
+If you cannot immediately see from the type where it is being used, this log message also gives you a stacktrace that can be used to set breakpoints and find out more details in your IDE.
+
+
+## Apache Thrift (via Kryo)
+
+In addition to the variants above, Flink also allows you to [register other type serialization frameworks](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/custom_serializers.html#register-a-custom-serializer-for-your-flink-program) with Kryo. After adding the appropriate dependencies from the [documentation](https://ci.apache.org/projects/flink/flink-docs-release-1.10/dev/custom_serializers.html#register-a-custom-serializer-for-your-flink-program) (`com.twitter:chill-thrift` and `org.apache.thrift:libthrift`), you can use [Apache Thrift](https://thrift.apache.org/) like the following:
+
+```java
+env.getConfig().addDefaultKryoSerializer(MyCustomType.class, TBaseSerializer.class);
+```
+
+This only works if (generic types are not disabled and) `MyCustomType` is a Thrift-generated data type or otherwise, it will fail at runtime with an exception like this:
```

Comment:
```suggestion
This only works if generic types are not disabled and `MyCustomType` is a Thrift-generated data type or otherwise, it will fail at runtime with an exception like this:
```

Reply:
In that case, I'd also have to change the failure description because that only relates to non-Thrift types. I think, it works better like this anyway:
```suggestion
This only works if generic types are not disabled and `MyCustomType` is a Thrift-generated data type. If the data type is not generated by Thrift, Flink will fail at runtime with an exception like this:
```